content/4_Gold/String_Hashing.mdx

```diff
@@ -184,13 +182,172 @@ public static long get_hash(int a, int b) { // Endpoints of the substring
+		for (int j = 0; j < 26; j++) match &= freq_curr[j] == freq_target[j];
+		if (match) good.insert({poly1, poly2});
+	}
+
+	cout << good.size() << '\n';
+	return 0;
+}
+```
+
+</CPPSection>
+
+</LanguageSection>
+
+### Two Hashes
+
+**Time Complexity:** $\mathcal O((|N| + |H|) \log M)$
+
+An alternative solution without frequency tables would be to hash the substrings that we're trying to match. Since order doesn't matter, we need to modify our hash function slightly.
+
+In particular, instead of computing the **polynomial** hash of the substrings, compute the product $(P + s_1)(P + s_2) \cdots (P + s_k) \bmod M$ as the hash (again, using two modulos).
+This hash is nice because the relative order of the letters doesn't matter, as multiplication is commutative.
+
+Since this hash requires the modular inverse, there's an extra $\log M$ factor in the time complexity.
+
+Alternative hashes (e.g. computing the sum $(P + s_1)^2 + (P + s_2)^2 + \dots + (P + s_k)^2 \bmod M$) also work for other hashing problems, but the test cases are too strong for that to pass here.
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+typedef long long ll;
+using namespace std;
+
+const ll P = 9973, M1 = 1e9 + 9, M2 = 1e9 + 7;
+
+ll inv(ll base, ll MOD) {
+	ll ans = 1, expo = MOD - 2;
+	while (expo) {
+		if (expo & 1) ans = ans * base % MOD;
+		expo >>= 1;
+		base = base * base % MOD;
+	}
+	return ans;
+}
+
+string n, h;
+
+int main() {
+	cin.tie(0)->sync_with_stdio(0);
+	cin >> n >> h;
+	if (n.size() > h.size()) return cout << 0, 0;
+
+	set<pair<ll, ll>> good;
+	ll n_hsh1 = 1, h_hsh1 = 1, n_hsh2 = 1, h_hsh2 = 1;
+	ll poly1 = 0, poly2 = 0, p_pow1 = 1, p_pow2 = 1;
+	for (int i = 0; i < n.size(); i++) {
+		// Compute product hashes
+		n_hsh1 = n_hsh1 * (P + n[i] - 'a') % M1;
+		n_hsh2 = n_hsh2 * (P + n[i] - 'a') % M2;
+		h_hsh1 = h_hsh1 * (P + h[i] - 'a') % M1;
+		h_hsh2 = h_hsh2 * (P + h[i] - 'a') % M2;
+		// Compute polynomial hashes
+		poly1 = (poly1 * P + (h[i] - 'a')) % M1;
+		poly2 = (poly2 * P + (h[i] - 'a')) % M2;
+		// p_pow1 and p_pow2 will be used to **update** the polynomial hashes
+		p_pow1 = p_pow1 * P % M1;
+		p_pow2 = p_pow2 * P % M2;
+	}
+	if (n_hsh1 == h_hsh1 && n_hsh2 == h_hsh2) good.insert({poly1, poly2});
+
+	for (int i = n.size(); i < h.size(); i++) {
+		// Update product hashes using modular inverse
+		h_hsh1 = h_hsh1 * inv(P + h[i - n.size()] - 'a', M1) % M1 * (P + h[i] - 'a') % M1;
+		h_hsh2 = h_hsh2 * inv(P + h[i - n.size()] - 'a', M2) % M2 * (P + h[i] - 'a') % M2;
+		// Update polynomial hashes
+		poly1 = (poly1 * P - p_pow1 * (h[i - n.size()] - 'a') % M1 + (h[i] - 'a') + M1) % M1;
+		poly2 = (poly2 * P - p_pow2 * (h[i - n.size()] - 'a') % M2 + (h[i] - 'a') + M2) % M2;
+		
+		if (n_hsh1 == h_hsh1 && n_hsh2 == h_hsh2) good.insert({poly1, poly2});
+	}
+
+	cout << good.size() << '\n';
+	return 0;
+}
+```
+
+</CPPSection>
+
+</LanguageSection>
+
+<Optional title="Probability of Collisions + Hashing a Tree">
```

Comment:
I guess there isn't actually a need to mention this since it's already linked in the blog linked above.